### PR TITLE
Programify feature gate

### PR DESCRIPTION
**Pull request by buffalojoec**
This is SIMD 1/3 expected for **Multi-Client Feature Gates**. See <https://github.com/solana-foundation/solana-improvement-documents/issues/76>

**Goals:**

- Decentralized control over queuing new runtime features
- Automatic feature activation selection based on stake weight with supporting
  software
- Decentralized governance of the mechanism itself

**Resulting Architecture:**

1. 👉 **Feature Creation:** Features can be created by anyone. Each is owned by
  an upgradeable BPF program at `Feature111111111111111111111111111111111111`
2. **Feature Queuing:** A governance process nominates features that should be
  queued for activation
3. **Feature Recognition & Activation:** Features are activated based on stake
  support of nodes who recognize the feature in their software version

## Summary

This SIMD outlines a proposal to replace the non-existent system account at
address `Feature111111111111111111111111111111111111`, which is the owner of
all feature accounts, with an upgradeable BPF program.

It defines the program's initial functionality - which consists solely of the
capability to revoke pending feature activations - and an accompanying
governance system for managing upgrades of the program.

Important note: the process by which core contributors *activate* features
would remain completely unchanged by this SIMD.

### CODE REVIEW

File: proposals/0077-programify-feature-gate.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+It defines the program's initial functionality - which consists solely of the
+capability to revoke pending feature activations - and an accompanying
+governance system for managing upgrades of the program.
+
+Important note: the process by which core contributors *activate* features
+would remain completely unchanged by this SIMD.
+
+## Motivation
+
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md:51]: I think I would start the Motivation section by discussing why we want a Feature-Gate BPF program: namely, that the complete Multi-Client Feature Gate architecture will require several changes to how feature accounts are handled, and the most obvious approach is to put a program in charge of this account handling. A BPF program (as opposed to a native one, comment on this to come later) makes sense because it comes with a built in system for upgrades.

File: proposals/0077-programify-feature-gate.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+Such a program surely presents a security risk if the upgrade authority is not
+carefully handled, so a proper governance process is required for this program
+to exist.
+
+## Alternatives Considered
+
+At this time, core contributors have not considered any viable alternative
+solutions for enabling the ability to revoke pending feature activations. An
+on-chain program was decided to be the optimally viable solution. Suggestions
+are welcome.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]: Instead of this paragraph, I would mention the alternative of making the Feature Gate program into a fully fledged native program, with the big drawback that any changes to the program would need to be implemented by all validator clients in coordination.

File: proposals/0077-programify-feature-gate.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+used as a special address that is known to the Solana SDK. After this process
+completes, the feature is scheduled for activation and will be activated at the
+next epoch boundary. 
+
+### Deploying the Program
+
+In order to deploy an upgradeable BPF program to the address at
+`Feature111111111111111111111111111111111111`, a runtime change is required.
+This change would allow the runtime, upon feature activation, to move an
+already deployed upgradeable BPF program in place of the non-existent
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
already deployed upgradeable BPF program into the account at address
```

File: proposals/0077-programify-feature-gate.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+### Proposal
+
+This SIMD outlines a proposal to replace the non-existent system account at
+address `Feature111111111111111111111111111111111111`, which is the owner of
+all feature accounts, with an upgradeable BPF program.
+
+It defines the program's initial functionality - which consists solely of the
+capability to revoke pending feature activations - and an accompanying
+governance system for managing upgrades of the program.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
capability to revoke pending feature activations - and an initial
governance system for managing upgrades of the program.
```
I think we should say this is just the starting approach, as validators may wish to control this with formal governance at some point in the future.
Also, I feel like the phrase "governance system" might imply something broader than a simple multisig. Is there a different phrase we could use here?

File: proposals/0077-programify-feature-gate.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+paramount to secure, frictionless engineering processes and protecting the
+network’s stability.
+
+The best way to manage such a capability is to ensure all feature accounts are
+owned by a program, which has the authority to clear a feature account’s data
+and remove its lamports, effectively revoking it from activation.
+
+Such a program surely presents a security risk if the upgrade authority is not
+carefully handled, so a proper governance process is required for this program
+to exist.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]: I thought this section could maybe bridge the BPF-program motivation I mentioned above and the "current state of things" you include in the design section below. What do you think? (Needs line wrapping)

```suggestion
Currently, a feature is queued for activation by a keypair holder creating an empty account and assigning it to the `Feature111111111111111111111111111111111111` program. Because there is no actual program at that address, the queuing is irreversible; if the runtime knows about a feature gate at that address, it will activate it at the next epoch boundary. This means there is no recourse in the case of a mistaken queuing, discovery of a bug in the code wrapped by the feature-gate, or just a desire to manage the cadence and schedule of activations. A fully-fledged BPF program would take ownership of those accounts and support revoking queued features, giving engineers more flexibility and safeguards.
```

File: proposals/0077-programify-feature-gate.md [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+authorized the activation, the feature can be activated using the keypair and
+the CLI command `solana feature activate`. 
+
+This command will use the keypair to authorize a transaction that will allocate
+the necessary data and assign it under owner
+`Feature111111111111111111111111111111111111`. This account at
+`Feature111111111111111111111111111111111111` does not exist, but is instead
+used as a special address that is known to the Solana SDK. After this process
+completes, the feature is scheduled for activation and will be activated at the
+next epoch boundary.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]: I prefer a short summary in Motivation instead of this section. Also, since SIMDs are supposed to be implementation agnostic, we shouldn't include links to labs client code. The SIMD needs to stand alone.

File: proposals/0077-programify-feature-gate.md [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+In order to deploy an upgradeable BPF program to the address at
+`Feature111111111111111111111111111111111111`, a runtime change is required.
+This change would allow the runtime, upon feature activation, to move an
+already deployed upgradeable BPF program in place of the non-existent
+`Feature111111111111111111111111111111111111`.
+
+For maximum security, the intial program that will be deployed and then moved
+in place of `Feature111111111111111111111111111111111111` will be a no-op
+program with an intentionally large allocation (to allow for larger programs
+in the future).
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion

For maximum security, the initial program can be a no-op
program with an intentionally large allocation (to allow for larger programs
in the future).
```

File: proposals/0077-programify-feature-gate.md [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+Shortly after verifying this program was successfully moved into place, the
+program will be upgraded to contain Feature Gate Program v0.1.0 - which will
+include the "revoke functionality defined below.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
Once the program is moved into place, the program can be upgraded using the conventional BpfUpgradeableLoader method to include the revoke functionality defined below.
```

File: proposals/0077-programify-feature-gate.md [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+<https://github.com/buffalojoec/noop-programs/tree/main/programs/feature-gate-noop>
+
+The runtime change to perform this account replacement can be found in the pull
+request linked below.
+
+<https://github.com/solana-labs/solana/pull/32783>
+
+The feature tracking issue for this change can be found here:
+
+<https://github.com/solana-labs/solana/issues/33547>
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]: Remove this stuff

File: proposals/0077-programify-feature-gate.md [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+
+### Revoking Pending Features
+
+As mentioned above under “New Terminology”, the act of revoking a pending
+feature activation consists of reallocating a feature account’s data to zero,
+assigning it to the system program, and defunding its lamports balance. This
+causes the feature account to be “revoked” since the runtime will no longer
+detect it as an account owned by `Feature11111`.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
detect it as an account owned by `Feature111111111111111111111111111111111111`.
```

Comment (CriesofCarrots) [reply proposals/0077-programify-feature-gate.md]: If you don't want to use the whole address everywhere, you could define the "Feature Program ID" somewhere up above?

File: proposals/0077-programify-feature-gate.md [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+
+When a core contributor executes the `solana feature activate` command, a
+signature from the feature key pair is required to activate it, since its state
+will change. Similarly, we can require the same feature key pair’s signature to
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
signature from the feature keypair is required to activate it, since its state
will change. Similarly, we can require the same feature keypair’s signature to
```

File: proposals/0077-programify-feature-gate.md [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+
+Shortly after verifying this program was successfully moved into place, the
+program will be upgraded to contain Feature Gate Program v0.1.0 - which will
+include the "revoke functionality defined below.
+
+The specific no-op program that will be initially deployed and moved by the
+runtime has been verifiably built and deployed to devent, testnet, and
+mainnet-beta. It can be found in the repository linked below.
+
+<https://github.com/buffalojoec/noop-programs/tree/main/programs/feature-gate-noop>
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
The specific no-op program to be initially moved into `Feature111111111111111111111111111111111111` by the
runtime should be verifiably built and then deployed to devent, testnet, and
mainnet-beta.
```

Comment (CriesofCarrots) [reply proposals/0077-programify-feature-gate.md]: Before we activate the feature gate to do the switch we can publish your repo link for examination by labs/jito/firedancer client teams and anyone else who wants to verify it.
But the SIMD is supposed to be generic.

File: proposals/0077-programify-feature-gate.md [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    .ok_or::<ProgramError>(FeatureGateError::Overflow.into())?;
+
+**feature_info.try_borrow_mut_lamports()? = 0;
+**destination_info.try_borrow_mut_lamports()? = new_destination_lamports;
+
+feature_info.realloc(0, true)?;
+feature_info.assign(&system_program::id());
+```
+
+### Governing the Program Upgrades
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
### Controlling Program Upgrades
```

File: proposals/0077-programify-feature-gate.md [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+
+Upgrading the Feature Gate program will be controlled by a 2-of-3
+multi-signature authority, consisting of key-holders distributed across network
+participants.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
Because the Feature Gate program requires special runtime support, upgrading the Feature Gate program will initially be controlled by a 2-of-3
multi-signature authority, consisting of key-holders distributed across validator-client
teams.
```

File: proposals/0077-programify-feature-gate.md [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+
+- Solana Labs: 1 key-holder
+- Jump: 1 key-holder
+- Jito: 1 key-holder
+
+Only when 2 out of 3 key-holders have authorized an upgrade will the Feature
+Gate program be upgraded.
+
+**Note:** This includes the upgrade required to upgrade the initial no-op
+program to Feature Gate Program v0.1.0.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
**Note:** This includes the upgrade required to upgrade the initial no-op
program to the first released Feature Gate Program supporting revoke.
```

File: proposals/0077-programify-feature-gate.md [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+
+## Impact
+
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md:198]: Call out no change to activation process or timing?

File: proposals/0077-programify-feature-gate.md [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+Core contributors are positively impacted by this change, since the ability to
+revoke pending feature activations is a significant security advantage.
+
+However, anyone in the community - validators, developers, etc. - who *is not*
+satisfied with the governance system outlined in this proposal may be
+emotionally impacted and/or feel there is a risk of centralized authority
+around a piece of vital runtime functionality.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]: Maybe rephrase to say that this SIMD provides more decentralized control over how feature gates work, but could be more decentralized in the future.

File: proposals/0077-programify-feature-gate.md [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+confidently update their state upon activation.
+
+With this proposal, a live BPF program - which can accept instructions from
+anyone and execute code - will be the owner of these accounts. Surely this
+provides significant risk if *both* the program’s processor code *and* its
+upgrade authority are not properly managed. 
+
+With that being said, thoroughly reviewed and safe processor code, as well as
+a robust governance system for upgrading the program, will together mitigate
+these new risks as much as possible.
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
Currently the accounts used for feature-gating are owned by an program id that
does not exist. This means that there’s no on-chain authority that can modify
feature accounts once they’ve been created. This allows the runtime to
confidently update their state upon activation.

With this proposal, a live BPF program - which can accept instructions from
anyone and execute code - will be the owner of these accounts. This
does create some risk if *both* the program’s processor code *and* its
upgrade authority are not properly managed.  But thoroughly reviewed and safe processor code, as well as
a decentralized system for upgrading the program, will together mitigate
these new risks as much as possible.
```

**Review by CriesofCarrots — COMMENTED** [review]
Nice! A few suggestions (none of which are character-wrapped, sorry... if you know of some good way to do that, short of c+p-ing out to my text editor, please let me know!).

File: proposals/0077-programify-feature-gate.md [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+
+This is SIMD 1/3 required for **Multi-Client Feature Gates**. See
+<https://github.com/solana-foundation/solana-improvement-documents/issues/76>
+
+**Goals:**
+
+- Decentralized control over queuing new runtime features
+- Automatic feature activation selection based on stake weight with supporting
+  software
+- Decentralized governance of the mechanism itself
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
- Decentralized control of the mechanism itself
```
Maybe? Just because governance kind of means a particular thing, in the crypto worls

File: proposals/0077-programify-feature-gate.md [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+status: Draft
+created: 2023-10-26
+feature: #33547
+---
+
+## Summary
+
+### Roadmap
+
+This is SIMD 1/3 required for **Multi-Client Feature Gates**. See
```
Comment (CriesofCarrots) [inline proposals/0077-programify-feature-gate.md]:
```suggestion
This is SIMD 1/3 expected for **Multi-Client Feature Gates**. See
```

File: proposals/0077-programify-feature-gate.md (hunk 4 above)
Comment (buffalojoec) [reply proposals/0077-programify-feature-gate.md]: Upgrade control process?

File: proposals/0077-programify-feature-gate.md (hunk 10 above)
Comment (buffalojoec) [reply proposals/0077-programify-feature-gate.md]: I tried this out but it honestly reads better with the full address.

File: proposals/0077-programify-feature-gate.md (hunk 4 above)
Comment (buffalojoec) [reply proposals/0077-programify-feature-gate.md]: I've replaced "governance system" with "upgrade control process" for now.

File: proposals/0077-programify-feature-gate.md (hunk 5 above)
Comment (buffalojoec) [reply proposals/0077-programify-feature-gate.md]: Nice! I like it!

**Comment by buffalojoec** [comment]
Thanks for the review! I broke the revisions up into a few commits. Once we're happy with it, I'll squash and open the SIMD for review.

**Review by CriesofCarrots — APPROVED** [review]
Awesome, thanks for the cleanup! I'm happy with this as a starting place for the SIMD
